specifications/LearnProofs/FindHighest.tla

```diff
@@ -0,0 +1,243 @@
+        <4>b. h <= h' BY <3>a DEFS lb, max
+        <4>c. \A idx \in 1..i : f[idx] <= h'
+          BY <4>a, <4>b DEF InductiveInvariant
+        <4>d. i = i' - 1 BY <3>a DEF lb
+        <4>e. UNCHANGED f
+          BY DEF lb
+        \* These steps are annotated to use the Zenon prover, which succeeds
+        \* immediately; otherwise we have to wait for other solvers to have
+        \* their turn and fail.
+        <4>f. \A idx \in 1..(i' - 1) : f'[idx] <= h'
+          BY Zenon, <4>c, <4>d, <4>e
+        <4> QED
+          BY Zenon, <4>f DEF InductiveInvariant
+      <3>b. CASE ~(i <= Len(f))
+        <4> UNCHANGED <<f, h, i>> BY <3>b DEF lb
+        <4> QED BY DEF InductiveInvariant
+      <3> QED BY <3>a, <3>b DEF lb
+    <2> QED BY <2>a, <2>b DEF Next
+  \* We need to note we made use of the type invariant theorem here
+  <1> QED BY PTL, <1>a, <1>b, <1>c, TypeInvariantHolds DEF Spec
+
+\* A small sub-theorem that relates our inductive invariant to correctness
+DoneIndexValue == pc = "Done" => i = Len(f) + 1
+
+THEOREM DoneIndexValueThm == Spec => []DoneIndexValue
+PROOF
+  <1>a. Init => DoneIndexValue
+    BY DEF Init, DoneIndexValue
+  <1>b. DoneIndexValue /\ UNCHANGED vars => DoneIndexValue'
+    BY DEFS DoneIndexValue, vars
+  <1>c. DoneIndexValue /\ TypeOK /\ Next => DoneIndexValue'
+    <2>a. DoneIndexValue /\ Terminating => DoneIndexValue'
+      BY DEFS DoneIndexValue, Terminating, vars
+    <2>b. DoneIndexValue /\ TypeOK /\ lb => DoneIndexValue'
+      <3> SUFFICES  ASSUME DoneIndexValue, TypeOK, lb
+                    PROVE DoneIndexValue'
+          OBVIOUS
+      <3> USE DEFS DoneIndexValue, TypeOK, lb
+      <3>a. CASE i <= Len(f)
+        <4>a. pc' /= "Done" BY <3>a
+        <4> QED BY <3>a, <4>a DEF lb
+      <3>b. CASE ~(i <= Len(f))
+        <4>b. i \in 1..(Len(f) + 1) BY DEF TypeOK
+        <4>c. i = Len(f) + 1 BY <3>b, <4>b
+        <4>d. UNCHANGED <<f, i>> BY <3>b
+        <4>e. i' = Len(f') + 1 BY <4>c, <4>d
+        <4>f. pc' = "Done" BY <3>b
+        <4> QED BY <4>e, <4>f DEF DoneIndexValue
+      <3> QED BY <3>a, <3>b DEF lb
+    <2> QED BY <2>a, <2>b DEF Next
+  <1> QED BY PTL, <1>a, <1>b, <1>c, TypeInvariantHolds DEF Spec
+
+\* The main event! After the algorithm has terminated, the variable h must
+\* have value greater than or equal to any element of the sequence.
+Correctness ==
+  pc = "Done" =>
+    \A idx \in DOMAIN f : f[idx] <= h
+
+THEOREM IsCorrect == Spec => []Correctness
+PROOF
+  <1>a. Init => Correctness
+    BY DEF Init, Correctness
+  <1>b. Correctness /\ UNCHANGED vars => Correctness'
+    BY DEF Correctness, vars
+  <1>c. /\ Correctness
+        /\ InductiveInvariant'
+        /\ DoneIndexValue'
+        /\ Next
+        => Correctness'
+    <2>a. Correctness /\ Terminating => Correctness'
+      BY DEF Correctness, Terminating, vars
+    <2>b.
+        /\ Correctness
+        /\ InductiveInvariant'
+        /\ DoneIndexValue'
+        /\ lb
+        => Correctness'
```

Comment:
Here as well, the following simple proof works:
       BY DEF Correctness, InductiveInvariant, DoneIndexValue, lb